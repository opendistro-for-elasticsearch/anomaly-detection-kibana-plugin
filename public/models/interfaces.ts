/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { DATA_TYPES, DETECTOR_STATE } from '../utils/constants';

export type FieldInfo = {
  label: string;
  type: DATA_TYPES;
};

export enum FILTER_TYPES {
  SIMPLE = 'simple_filter',
  CUSTOM = 'custom_filter',
}

export enum FEATURE_TYPE {
  SIMPLE = 'simple_aggs',
  CUSTOM = 'custom_aggs',
}

export enum OPERATORS_MAP {
  IS = 'is',
  IS_NOT = 'is_not',
  IS_NULL = 'is_null',
  IS_NOT_NULL = 'is_not_null',
  IS_GREATER = 'is_greater',
  IS_GREATER_EQUAL = 'is_greater_equal',
  IS_LESS = 'is_less',
  IS_LESS_EQUAL = 'is_less_equal',
  STARTS_WITH = 'starts_with',
  ENDS_WITH = 'ends_with',
  CONTAINS = 'contains',
  NOT_CONTAINS = 'does_not_contains',
  IN_RANGE = 'in_range',
  NOT_IN_RANGE = 'not_in_range',
}

export type UIFilter = {
  fieldInfo: FieldInfo[];
  operator: OPERATORS_MAP;
  fieldValue?: string | number | null;
  fieldRangeStart?: number;
  fieldRangeEnd?: number;
};

export type FeatureAttributes = {
  featureId?: string;
  featureName: string;
  featureEnabled: boolean;
  importance: number;
  aggregationQuery: { [key: string]: any };
};

export enum UNITS {
  MINUTES = 'MINUTES',
}

export type Schedule = {
  interval: number;
  unit: UNITS;
};

export type UiFeature = {
  featureType: FEATURE_TYPE;
  aggregationBy?: string;
  aggregationOf?: string;
};

export type UiMetaData = {
  filterType: FILTER_TYPES;
  filters: UIFilter[];
  features: {
    [key: string]: UiFeature;
  };
};
export type Detector = {
  primaryTerm: number;
  seqNo: number;
  id: string;
  name: string;
  description: string;
  timeField: string;
  indices: string[];
  filterQuery: { [key: string]: any };
  featureAttributes: FeatureAttributes[];
  windowDelay: { period: Schedule };
  detectionInterval: { period: Schedule };
  uiMetadata: UiMetaData;
  lastUpdateTime: number;
  enabled?: boolean;
  enabledTime?: number;
  disabledTime?: number;
  curState: DETECTOR_STATE;
  initializationError: string;
};

export type DetectorListItem = {
  id: string;
  name: string;
  indices: string[];
  curState: DETECTOR_STATE;
  featureAttributes: FeatureAttributes[];
  totalAnomalies: number;
  lastActiveAnomaly: number;
  lastUpdateTime: number;
};

export type AnomalyData = {
  anomalyGrade: number;
  anomalyScore?: number;
  confidence: number;
  detectorId?: string;
  endTime: number;
  startTime: number;
  plotTime?: number;
};

export type FeatureAggregationData = {
  data: number;
  endTime: number;
  startTime: number;
  plotTime?: number;
};

export type Anomalies = {
  anomalies: AnomalyData[];
  featureData: { [key: string]: FeatureAggregationData[] };
};

export type Monitor = {
  id: string;
  name: string;
  enabled: boolean;
  enabledTime?: number;
  schedule: { period: Schedule };
  inputs: any[];
  triggers: any[];
  lastUpdateTime: number;
};

export type MonitorAlert = {
  monitorName: string;
  triggerName: string;
  severity: number;
  state: string;
  error: string;
  startTime: number;
  endTime: number;
  acknowledgedTime: number;
};

export type AnomalySummary = {
  anomalyOccurrence: number;
  minAnomalyGrade: number;
  maxAnomalyGrade: number;
  minConfidence: number;
  maxConfidence: number;
  lastAnomalyOccurrence: string;
};

export type DateRange = {
  startDate: number;
  endDate: number;
};
