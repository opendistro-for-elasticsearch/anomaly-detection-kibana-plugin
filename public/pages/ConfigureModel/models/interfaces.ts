/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { FEATURE_TYPE } from '../../../models/interfaces';
import { AggregationOption } from '../../../models/types';

// Formik values used when creating/editing the model configuration
export interface ModelConfigurationFormikValues {
  featureList: FeaturesFormikValues[];
  categoryFieldEnabled: boolean;
  categoryField: string[];
  shingleSize: number;
}

export interface FeaturesFormikValues {
  featureId: string;
  featureName: string | undefined;
  featureType: FEATURE_TYPE;
  featureEnabled: boolean;
  aggregationQuery: string;
  aggregationBy?: string;
  aggregationOf?: AggregationOption[];
  newFeature?: boolean;
}
