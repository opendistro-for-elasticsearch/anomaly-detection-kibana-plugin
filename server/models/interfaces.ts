/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export interface DefaultHeaders {
  'Content-Type': 'application/json';
  Accept: 'application/json';
  'User-Agent': 'Kibana';
}

export interface SearchResponse<T> {
  hits: {
    total: { value: number };
    hits: Array<{
      _source: T;
      _id: string;
      _seq_no?: number;
      _primary_term?: number;
    }>;
  };
}

export interface ADApis {
  [API_ROUTE: string]: string;
  readonly DETECTOR_BASE: string;
}

export interface AlertingApis {
  [API_ROUTE: string]: string;
  readonly ALERTING_BASE: string;
}
export interface Entity {
  name: string;
  value: string;
}
export interface Anomaly {
  anomalyGrade: number;
  confidence: number;
  anomalyScore: number;
  startTime: number;
  endTime: number;
  plotTime: number;
  entity?: Entity[];
}
// Plot time is middle of start and end time to provide better visualization to customers
// Example, if window is 10 mins, in a given startTime and endTime of 12:10 to 12:20 respectively.
// plotTime will be 12:15.
export interface FeatureData {
  startTime: number;
  endTime: number;
  plotTime: number;
  data: number;
}
export interface AnomalyResults {
  anomalies: Anomaly[];
  featureData: { [key: string]: FeatureData[] };
}

export interface InitProgress {
  percentageStr: string;
  estimatedMinutesLeft: number;
  neededShingles: number;
}

export interface EntityAnomalySummary {
  startTime: number;
  maxAnomaly: number;
  anomalyCount: number;
}

export interface EntityAnomalySummaries {
  entity: Entity;
  anomalySummaries: EntityAnomalySummary[];
}
