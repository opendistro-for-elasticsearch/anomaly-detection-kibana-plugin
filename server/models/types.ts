/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { SORT_DIRECTION } from '../utils/constants';

export type CatIndex = {
  index: string;
  health: string;
};

export type IndexAlias = {
  index: string;
  alias: string;
};

export type GetAliasesResponse = {
  aliases: IndexAlias[];
};

export type GetIndicesResponse = {
  indices: CatIndex[];
};

export type GetMappingResponse = {
  [key: string]: any;
};

export enum UNITS {
  MINUTES = 'MINUTES',
}

export type Schedule = {
  interval: number;
  unit: UNITS;
};

export type FeatureAttributes = {
  featureId?: string;
  featureName: string;
  featureEnabled: boolean;
};

export type Detector = {
  id?: string;
  name: string;
  description: string;
  indices: string[];
  filterQuery?: { [key: string]: any };
  featureAttributes?: FeatureAttributes[];
  windowDelay?: { period: Schedule };
  detectionInterval?: { period: Schedule };
  uiMetadata?: { [key: string]: any };
  lastUpdateTime: number;
  enabled: boolean;
  enabledTime?: number;
  disabledTime?: number;
};

export type Monitor = {
  id: string;
  name: string;
  enabled: boolean;
  enabledTime?: number;
  schedule: { period: Schedule };
  inputs: any[];
  triggers: any[];
  lastUpdateTime: number;
};

export type GetDetectorsQueryParams = {
  from: number;
  size: number;
  search: string;
  indices?: string;
  sortDirection: SORT_DIRECTION;
  sortField: string;
};

export type GetAdMonitorsQueryParams = {
  from: number;
  size: number;
  search: string;
  indices?: string;
  sortDirection: SORT_DIRECTION;
  sortField: string;
};

export type DetectorResultsQueryParams = {
  from: number;
  size: number;
  sortDirection: SORT_DIRECTION;
  sortField: string;
  dateRangeFilter?: DateRangeFilter;
};

export type AnomalyResult = {
  startTime: number;
  endTime: number;
  plotTime: number;
  anomalyGrade: number;
  // grade: number;
  confidence: number;
};

export type FeatureResult = {
  startTime: number;
  endTime: number;
  plotTime: number;
  data: number;
};

export type AnomalyResultsResponse = {
  totalAnomalies: number;
  results: AnomalyResult[];
  featureResults: { [key: string]: FeatureResult[] };
};

export type ServerResponse<T> =
  | { ok: false; error: string }
  | { ok: true; response: T };

export type DateRangeFilter = {
  startTime?: number;
  endTime?: number;
  fieldName: string;
}