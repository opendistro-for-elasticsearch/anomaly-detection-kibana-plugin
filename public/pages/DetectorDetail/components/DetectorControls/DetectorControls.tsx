/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { useState } from 'react';
import {
  EuiButton,
  EuiContextMenuItem,
  EuiContextMenuPanel,
  EuiFlexGroup,
  EuiFlexItem,
  EuiPopover,
} from '@elastic/eui';
import { Detector } from '../../../../models/interfaces';

interface DetectorControls {
  onEditDetector(): void;
  onEditFeatures(): void;
  onDelete(): void;
  onStartDetector(): void;
  onStopDetector(): void;
  detector: Detector;
}
export const DetectorControls = (props: DetectorControls) => {
  const [isOpen, setIsOpen] = useState(false);
  return (
    <EuiFlexGroup justifyContent="spaceBetween" alignItems="center">
      <EuiFlexItem grow={false}>
        <EuiPopover
          id="actionsPopover"
          button={
            <EuiButton
              iconType="arrowDown"
              iconSide="right"
              data-test-subj="actionsButton"
              onClick={() => setIsOpen(!isOpen)}
            >
              Actions
            </EuiButton>
          }
          panelPaddingSize="none"
          anchorPosition="downLeft"
          isOpen={isOpen}
          closePopover={() => setIsOpen(false)}
        >
          <EuiContextMenuPanel>
            <EuiContextMenuItem
              key="edit"
              data-test-subj="editDetector"
              onClick={props.onEditDetector}
            >
              Edit detector
            </EuiContextMenuItem>

            <EuiContextMenuItem
              key="edit"
              data-test-subj="editFeature"
              onClick={props.onEditFeatures}
            >
              Edit features
            </EuiContextMenuItem>

            <EuiContextMenuItem
              key="deleteDetector"
              data-test-subj="deleteDetector"
              onClick={props.onDelete}
            >
              Delete detector
            </EuiContextMenuItem>
          </EuiContextMenuPanel>
        </EuiPopover>
      </EuiFlexItem>
      <EuiFlexItem grow={false}>
        <EuiButton
          data-test-subj="stopDetectorButton"
          onClick={
            props.detector.enabled
              ? props.onStopDetector
              : props.onStartDetector
          }
          iconType={props.detector.enabled ? 'stop' : 'play'}
          disabled={
            !props.detector.featureAttributes ||
            props.detector.featureAttributes.length === 0
          }
        >
          {props.detector.enabled ? 'Stop detector' : 'Start detector'}
        </EuiButton>
      </EuiFlexItem>
    </EuiFlexGroup>
  );
};
