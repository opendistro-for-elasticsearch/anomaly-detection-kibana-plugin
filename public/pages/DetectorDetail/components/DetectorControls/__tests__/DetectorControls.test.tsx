/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { render } from '@testing-library/react';
import { DetectorControls } from '../DetectorControls';
import { UNITS } from '../../../../../models/interfaces';

describe('<DetectorControls /> spec', () => {
  const detector = {
    primaryTerm: 1,
    seqNo: 1,
    id: '',
    name: '4QY4YHEB5W9C7vlb3Mou',
    description: 'test detector',
    timeField: 'timestamp',
    indices: ['test-index'],
    filterQuery: { match_all: { boost: 1.0 } },
    featureAttributes: [],
    windowDelay: { period: { interval: 0, unit: UNITS.MINUTES } },
    detectionInterval: { period: { interval: 0, unit: UNITS.MINUTES } },
    uiMetadata: {},
    lastUpdateTime: 1586714882000,
    enabled: false,
    enabledTime: 1586722082000,
  };
  const features = [
    {
      featureId: 'ZNU5YHEBTQyq-K19pyEP',
      featureName: 'feature1',
      featureEnabled: true,
      importance: 1,
      aggregationQuery: {
        feature1: { sum: { field: 'value' } },
      },
    },
  ];
  const onEditDetector = jest.fn();
  const onEditFeatures = jest.fn();
  const onDelete = jest.fn();
  const onStartDetector = jest.fn();
  const onStopDetector = jest.fn();
  describe('Detector Controls', () => {
    test('renders detector controls with disabled detecor and empty features', () => {
      const { container } = render(
        <DetectorControls
          onEditDetector={onEditDetector}
          onEditFeatures={onEditFeatures}
          onDelete={onDelete}
          onStartDetector={onStartDetector}
          onStopDetector={onStopDetector}
          // @ts-ignore
          detector={detector}
        />
      );
      expect(container).toMatchSnapshot();
    });

    test('renders detector controls with enabled detecor and empty features', () => {
      const enabledDetector = { ...detector, enabled: true };
      const { container } = render(
        <DetectorControls
          onEditDetector={onEditDetector}
          onEditFeatures={onEditFeatures}
          onDelete={onDelete}
          onStartDetector={onStartDetector}
          onStopDetector={onStopDetector}
          // @ts-ignore
          detector={enabledDetector}
        />
      );
      expect(container).toMatchSnapshot();
    });

    test('renders detector controls with disabled detecor and features', () => {
      const disabledDetectorWithFeatures = {
        ...detector,
        featureAttributes: features,
      };
      const { container } = render(
        <DetectorControls
          onEditDetector={onEditDetector}
          onEditFeatures={onEditFeatures}
          onDelete={onDelete}
          onStartDetector={onStartDetector}
          onStopDetector={onStopDetector}
          // @ts-ignore
          detector={disabledDetectorWithFeatures}
        />
      );
      expect(container).toMatchSnapshot();
    });

    test('renders detector controls with enabled detecor and features', () => {
      const disabledDetectorWithFeatures = {
        ...detector,
        enabled: true,
        featureAttributes: features,
      };
      const { container } = render(
        <DetectorControls
          onEditDetector={onEditDetector}
          onEditFeatures={onEditFeatures}
          onDelete={onDelete}
          onStartDetector={onStartDetector}
          onStopDetector={onStopDetector}
          // @ts-ignore
          detector={disabledDetectorWithFeatures}
        />
      );
      expect(container).toMatchSnapshot();
    });
  });
});
