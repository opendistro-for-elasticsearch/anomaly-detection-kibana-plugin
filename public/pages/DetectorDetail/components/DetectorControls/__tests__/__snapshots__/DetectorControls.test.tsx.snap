// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetectorControls /> spec Detector Controls renders detector controls with disabled detecor and empty features 1`] = `
<div>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
      style="margin-right: 16px;"
    >
      <div
        class="euiPopover euiPopover--anchorDownLeft"
        id="actionsPopover"
      >
        <div
          class="euiPopover__anchor"
        >
          <button
            class="euiButton euiButton--primary"
            data-test-subj="actionsButton"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonContent--iconRight euiButton__content"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonContent__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              />
              <span
                class="euiButton__text"
              >
                Actions
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
      style="margin-left: 0px;"
    >
      <button
        class="euiButton euiButton--primary euiButton-isDisabled"
        data-test-subj="stopDetectorButton"
        disabled=""
        type="button"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <svg
            aria-hidden="true"
            class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonContent__icon"
            focusable="false"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          />
          <span
            class="euiButton__text"
          >
            Start detector
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<DetectorControls /> spec Detector Controls renders detector controls with disabled detecor and features 1`] = `
<div>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
      style="margin-right: 16px;"
    >
      <div
        class="euiPopover euiPopover--anchorDownLeft"
        id="actionsPopover"
      >
        <div
          class="euiPopover__anchor"
        >
          <button
            class="euiButton euiButton--primary"
            data-test-subj="actionsButton"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonContent--iconRight euiButton__content"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiButtonContent__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13.069 5.157L8.384 9.768a.546.546 0 01-.768 0L2.93 5.158a.552.552 0 00-.771 0 .53.53 0 000 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 000-.76.552.552 0 00-.771 0z"
                  fill-rule="non-zero"
                />
              </svg>
              <span
                class="euiButton__text"
              >
                Actions
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
      style="margin-left: 0px;"
    >
      <button
        class="euiButton euiButton--primary"
        data-test-subj="stopDetectorButton"
        type="button"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <svg
            aria-hidden="true"
            class="euiIcon euiIcon--medium euiButtonContent__icon"
            focusable="false"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4.608 3.063C4.345 2.895 4 3.089 4 3.418v9.167c0 .329.345.523.608.356l7.2-4.584a.426.426 0 000-.711l-7.2-4.583zm.538-.844l7.2 4.583a1.426 1.426 0 010 2.399l-7.2 4.583C4.21 14.38 3 13.696 3 12.585V3.418C3 2.307 4.21 1.624 5.146 2.22z"
            />
          </svg>
          <span
            class="euiButton__text"
          >
            Start detector
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<DetectorControls /> spec Detector Controls renders detector controls with enabled detecor and empty features 1`] = `
<div>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
      style="margin-right: 16px;"
    >
      <div
        class="euiPopover euiPopover--anchorDownLeft"
        id="actionsPopover"
      >
        <div
          class="euiPopover__anchor"
        >
          <button
            class="euiButton euiButton--primary"
            data-test-subj="actionsButton"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonContent--iconRight euiButton__content"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiButtonContent__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13.069 5.157L8.384 9.768a.546.546 0 01-.768 0L2.93 5.158a.552.552 0 00-.771 0 .53.53 0 000 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 000-.76.552.552 0 00-.771 0z"
                  fill-rule="non-zero"
                />
              </svg>
              <span
                class="euiButton__text"
              >
                Actions
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
      style="margin-left: 0px;"
    >
      <button
        class="euiButton euiButton--primary euiButton-isDisabled"
        data-test-subj="stopDetectorButton"
        disabled=""
        type="button"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <svg
            aria-hidden="true"
            class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonContent__icon"
            focusable="false"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          />
          <span
            class="euiButton__text"
          >
            Stop detector
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<DetectorControls /> spec Detector Controls renders detector controls with enabled detecor and features 1`] = `
<div>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
      style="margin-right: 16px;"
    >
      <div
        class="euiPopover euiPopover--anchorDownLeft"
        id="actionsPopover"
      >
        <div
          class="euiPopover__anchor"
        >
          <button
            class="euiButton euiButton--primary"
            data-test-subj="actionsButton"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonContent--iconRight euiButton__content"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiButtonContent__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13.069 5.157L8.384 9.768a.546.546 0 01-.768 0L2.93 5.158a.552.552 0 00-.771 0 .53.53 0 000 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 000-.76.552.552 0 00-.771 0z"
                  fill-rule="non-zero"
                />
              </svg>
              <span
                class="euiButton__text"
              >
                Actions
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
      style="margin-left: 0px;"
    >
      <button
        class="euiButton euiButton--primary"
        data-test-subj="stopDetectorButton"
        type="button"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <svg
            aria-hidden="true"
            class="euiIcon euiIcon--medium euiButtonContent__icon"
            focusable="false"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 2h8a2 2 0 012 2v8a2 2 0 01-2 2H4a2 2 0 01-2-2V4a2 2 0 012-2zm0 1a1 1 0 00-1 1v8a1 1 0 001 1h8a1 1 0 001-1V4a1 1 0 00-1-1H4z"
              fill-rule="evenodd"
            />
          </svg>
          <span
            class="euiButton__text"
          >
            Stop detector
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
